src/analyses/ai_history.h

```diff
@@ -0,0 +1,237 @@
+/*******************************************************************\
+
+Module: Abstract Interpretation
+
+Author: Martin Brain, martin.brain@cs.ox.ac.uk
+
+\*******************************************************************/
+
+/// \file
+/// Abstract Interpretation history
+
+#ifndef CPROVER_ANALYSES_AI_HISTORY_H
+#define CPROVER_ANALYSES_AI_HISTORY_H
+
+#include <memory>
+
+#include <goto-programs/goto_model.h>
+
+#include <util/json.h>
+#include <util/xml.h>
+
+/// A history object is an abstraction / representation of the control
+/// control-flow part of a set of traces.  The simplest example is a single
+/// location which represents "all traces that reach this location".  More
+/// sophisticated history objects represent "all traces that reach this point
+/// after exactly N iterations of this loop", "all traces that follow a
+/// given pattern of branching to reach this point" or "all traces that have
+/// the following call stack and reach this location".
+///
+/// These are used to control the abstract interpreter; edges are computed
+/// per history.  Depending on what storage is used, they may also control
+/// or influence the number of domains that are used, supporting delayed
+/// merging, loop unwinding, context dependent analysis, etc.
+
+/// This is the base interface; derive from this.
+class ai_history_baset
+{
+public:
+  typedef goto_programt::const_targett locationt;
+  typedef ai_history_baset tracet;
+
+  /// History objects are intended to be immutable so they can be shared
+  /// to reduce memory overhead
+  typedef std::shared_ptr<const tracet> trace_ptrt;
+
+  /// Create a new history starting from a given location
+  /// This is used to start the analysis from scratch
+  /// PRECONDITION(l.is_dereferenceable());
+  explicit ai_history_baset(locationt)
+  {
+  }
+
+  ai_history_baset(const ai_history_baset &)
+  {
+  }
+
+  virtual ~ai_history_baset()
+  {
+  }
+
+  enum step_status
+  {
+    NEW,
+    MERGED,
+    BLOCKED
+  };
+
+  /// In a number of places we need to work with sets of histories
+  /// so that these (a) make use of the immutability and sharing and
+  /// (b) ownership can be transfered if necessary, we use shared pointers
+  /// rather than references.
+  /// The custom comparison allows particular histories to control
+  /// the order of exploration.
+  struct compare_historyt
+  {
+    bool operator()(const trace_ptrt &l, const trace_ptrt &r) const
+    {
+      return *l < *r;
+    }
+  };
+  typedef std::set<trace_ptrt, compare_historyt> trace_sett;
```

Comment:
I guess some people would prefer the `unordered` variant. With `shared_ptr` you'd get `std::hash` support for free.

Reply:
I would prefer to keep it ordered.  Histories have to have an ordering for working out which to propagate first in the abstract interpreter (see `working_sett`) and I would rather keep to just one kind of set.

Reply:
FWIW it's worth commenting if there was a deliberate choice to use an ordered map here, lest someone 'improve' it in the future (and it passing the regression tests by accident).

Reply:
I could write a test case that will break if the order of exploration / working_sett is changed substantially?  Would that help?

Reply:
@martin-cs The problem with tests like this is that if an unordered collection is used you can always get the ordering you _meant_ to get by sheer accident. This has bitten us in other places in the past. I think it's fair enough to just leave a comment here and hopefully that should at the very least flag things up during review

Reply:
I have expanded the comment to make it clearer why the ordering and per-history control of it is important.